fifeforspark/base_modelers.py

```diff
@@ -0,0 +1,430 @@
+        if self.weight_col:
+            self.reserved_cols.append(self.weight_col)
+        if self.data is not None:
+            self.categorical_features = [
+                col[0] for col in self.data.dtypes if col[1] == 'string']
+            self.numeric_features = [feature for feature in self.data.columns
+                                     if feature not in (self.categorical_features + self.reserved_cols)]
+            self.data = self.transform_features()
+
+    @abstractmethod
+    def train(self) -> Any:
+        """Train and return a model."""
+
+    @abstractmethod
+    def predict(
+        self, subset: Union[None, pyspark.sql.column.Column] = None, cumulative: bool = True
+    ) -> np.ndarray:
+        """Use trained model to produce observation survival probabilities."""
+
+    @abstractmethod
+    def evaluate(
+        self,
+        subset: Union[None, pyspark.sql.column.Column] = None,
+        threshold_positive: Union[None, str, float] = 0.5,
+        share_positive: Union[None, str, float] = None,
+    ) -> ks.frame.DataFrame:
+        """Tabulate model performance metrics."""
+
+    @abstractmethod
+    def forecast(self) -> pyspark.sql.DataFrame:
+        """Tabulate survival probabilities for most recent observations."""
+
+    @abstractmethod
+    def subset_for_training_horizon(
+        self, data: pyspark.sql.DataFrame, time_horizon: int
+    ) -> pyspark.sql.DataFrame:
+        """Return only observations where the outcome is observed."""
+
+    @abstractmethod
+    def label_data(self, time_horizon: int) -> pyspark.sql.DataFrame:
+        """Return data with an outcome label for each observation."""
+
+    @abstractmethod
+    def save_model(self, path: str = "") -> None:
+        """Save model file(s) to disk."""
+
+    @abstractmethod
+    def transform_features(self) -> pyspark.sql.DataFrame:
+        """Transform datetime features to suit model training."""
+
+    @abstractmethod
+    def build_model(self, n_intervals: Union[None, int] = None) -> None:
+        """Configure, train, and store a model."""
+
+    def set_n_intervals(self) -> int:
+        """Determine the maximum periods ahead the model will predict."""
+        train_durations = self.data.select(
+            when(
+                self.data[self.duration_col] <= self.data[self.max_lead_col], self.data[self.duration_col]
+            ).otherwise(self.data[self.max_lead_col]).alias('max_durations')
+        )
+        subset = ~self.data[self.validation_col] & ~self.data[self.test_col] & ~self.data[self.predict_col]
+        train_obs_by_lead_length = train_durations.filter(subset)
+        train_obs_by_lead_length = train_obs_by_lead_length.groupBy(
+            'max_durations').count()
+        n_intervals = train_obs_by_lead_length.filter(
+            train_obs_by_lead_length['count'] > self.config.get(
+                "min_survivors_in_train", 64).alias('max_durations')).agg(
+            {'max_durations': 'max'}
+        ).first()[0]
```

Comment:
I haven't thought about it enough, but if the filter returns an empty dataframe (not enough survivors in each period), this will crash without much of an explanation. Do we want to check first to see if n_intervals.first() is empty and if so raise an error?